### PR TITLE
Improve configuration to allow running code

### PR DESCRIPTION
**Pull request by LuigiPiucco**
## tl;dr
This allows running arbitrary user code _after_ Doom initialization, when modules are loaded and ready to `require`.
There's also discussion a perhaps better method, albeit one that would cause breaking changes.

## Full story
First of all, I'd like to say that this project is very interesting. I just changed from LunarVim, and I'm quite happy, but there were some obstacles I faced, namely, the apparent lack of support for just "running code" after Doom startup. Add to it that this is a very young project without all the options accounted for, and you run into issues with configuring plugins beyond a limit. That's where this PR comes in.

It implements a new configuration file, `doom_startup.lua`, that is run after everything else (see [`init.lua`](https://github.com/LuigiPiucco/doom-nvim/blob/develop/init.lua#L58-L63)), and really it just runs. This allows state modifications, such as: 

- Adding `which-key` bindings yourself;
- Overriding lsp config per language (this is needed in NixOS, for instance, where I actually want the carefully chosen server in `$PATH` to be used in place of whatever `lspinstall` downloads);
- Loading extensions which need to be registered, such as `Telescope` plugins;
- Really many more things.

Right now, the only limitation imposed on the startup file is that it returns to the caller a table with `source` set to its path, since all other config files also rely on that. This introduces a bit of boilerplate:

```lua
local M = {}

M.source = debug.getinfo(1, "S").source:sub(2)

-- Your startup logic...

return M
```

## Alternatives
In Doom Emacs, this functionality is done through `$DOOMDIR/config.el` and, in fact, there's no real equivalent of Doom Nvim's `doom_config.lua`, because settings are just variables that can be set, and "running code" includes setting variables. Therefore, maybe it would be better in the long run to change config behavior to more closely match that. An approach I can think of is similar to LunarVim's: inject a table into the config and allow the user to change it to reflect their taste. We would need one (maybe two, depending on the granularity wanted) special keys, something like `post_startup` and (optionally) `pre_startup`, that could be set to a function to be run after/before everything is loaded.

If requested, I could work on such changes, though I'm a little low on time for this and the next week, I may be slow to finish.

### CODE REVIEW

File: lua/doom/core/config/startup.lua [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+-- Check if running Neovim or Vim and fails if:
+-- 1. Running Vim instead of Neovim
+-- 2. Running Neovim 0.4 or below
+if vim.fn.has("nvim") == 1 then
+  if vim.fn.has("nvim-0.5") ~= 1 then
+    log.fatal("Doom Nvim requires Neovim 0.5.0, please update it")
+  end
+else
+  log.fatal("Doom Nvim does not have support for Vim, please use it with Neovim instead")
+end
```
Comment (NTBBloodbath) [inline lua/doom/core/config/startup.lua:15]: We are already checking this in `lua/doom/core/config/init.lua` IIRC, it actually makes more sense to me to move this code to this file as you did.

However, we would need to actually call this file at top before start loading Doom core files so it will raise the error at the correct time IMO. Up to you what to do tho, deleting this chunk from this file or deleting this from the file that I told you in order to avoid duplicated code and reordering how are Doom files being loaded 🤔

**Comment by NTBBloodbath** [comment]
Hi, thank you so much per doing this and also per your interest in this project :). The idea is pretty cool!

Some of the state modifications (the `which-key` bindings and NixOS lsp setup) are planned to be resolved in our core by using `nest.nvim` to map bindings and also adding some extra logic into lsp-installer module btw but that's another thing that should be discussed another day haha. I completely understand the use case of the new file and I think we really needed this to be a thing :p

> Right now, the only limitation imposed on the startup file is that it returns to the caller a table with source set to its path, since all other config files also rely on that. This introduces a bit of boilerplate

Yeah, sadly we cannot get rid of that boilerplate because it allows our configuration files to be easily backed up in dotfiles and loaded from literally anywhere (_also improves our integration with [cheovim](https://github.com/NTBBloodbath/cheovim)_).

> In Doom Emacs, this functionality is done through `$DOOMDIR/config.el` and, in fact, there's no real equivalent of Doom Nvim's `doom_config.lua`, because settings are just variables that can be set, and "running code" includes setting variables. Therefore, maybe it would be better in the long run to change config behavior to more closely match that.

Agreed, we still have a long way to be a 1:1 port of the Emacs framework but hey the project is still very young :p. We could start with your implementation in a separated file and progressively move the logic to our main config file by replacing some actual stuff.

> An approach I can think of is similar to LunarVim's: inject a table into the config and allow the user to change it to reflect their taste. We would need one (maybe two, depending on the granularity wanted) special keys, something like `post_startup` and (optionally) `pre_startup`, that could be set to a function to be run after/before everything is loaded.

This actually sounds great! I think we should go with two tables, granularity could be a very good option for us in this case 🙂

> If requested, I could work on such changes, though I'm a little low on time for this and the next week, I may be slow to finish.

Sure thing, do it if you feel that you can :). No rush so take your time, also don't worry about asking me to help you if you have some questions or need to understand something.

Cheers :)

File: lua/doom/core/config/startup.lua (hunk 1 above)
Comment (LuigiPiucco) [reply lua/doom/core/config/startup.lua:15]: For now, I think leaving it on `config/init.lua` is better, because we want to fail fast on unsupported systems and this file really needs to be loaded last. This is actually only here due to careless copy-and-paste, but maybe it should be in a more logical place, such as the root `init.lua`.  I will look into it later.

File: init.lua [hunk 2]
```diff
@@ -54,3 +54,10 @@ vim.defer_fn(function()
     ]])
   end
 end, 0)
+
+vim.defer_fn(function ()
```
[the remark this answers is not in the document]
Comment (LuigiPiucco) [reply init.lua:58]: Indeed, and looking above I see doom has a `file_exists` util, I can swap `filereadable` for that.

**Comment by LuigiPiucco** [comment]
Hello again, I have read your review and am going to chance the code as mentioned in my replies. I'm not close to the computer right now, but it should be ready by the end of the day (that's a UTC -3 day, which just started, though).

Also, thanks for being very receptive!

**Comment by LuigiPiucco** [comment]
I've applied the modifications discussed previously and rebased, plus there's an extra commit for documentation in `docs/getting_started.md`.

About the version check, I hoisted it to the root `init.lua`, there seems to be no problem, though I don't have nvim 0.4 easily available to properly check if it fails. Beyond that, note that it may be unneeded anyway, since versions prior to 0.5 of Neovim and all versions of Vim don't support lua, therefore giving errors on trying to load the file.

**Review by NTBBloodbath — APPROVED** [review]

**Comment by LuigiPiucco** [comment]
I forgot to require `doom.extras.log`. Oops!

About the changes to how configuration works proposed in the opening post, I started working on them, but I don't have enough to share right now, and may not have for the coming week. I will open an issue describing the prototype implementation once it has actual features (right now I just re-implemented the config to work with a global table instead of returning values). What I can say before hand is that this PR will be more-or-less obsolesced, since I plan to add the `post_startup` key I mentioned.

That being said, it may be some time until that, so I would recommend merging this while we still need it.

**Comment by LuigiPiucco** [comment]
This is superseded by what is being implemented for #233, I'll close this and explain it there later. Long story short, we'll have ways to add custom logic to packer configs through `doom_config.lua`, which removes the need for this.
